crates/ruff_python_semantic/src/scope.rs

```diff
@@ -19,6 +21,8 @@ pub struct Scope<'a> {
     bindings: FxHashMap<&'a str, BindingId>,
     /// A map from bound name to binding index, for bindings that were shadowed later in the scope.
     shadowed_bindings: FxHashMap<&'a str, Vec<BindingId>>,
+    /// A map from global name to the range that declares it.
+    globals: FxHashMap<&'a str, TextRange>,
```

Comment:
I actually have a branch that removes this, but it's a little less computationally efficient. We have to store these upfront so that we can detect `Expr::Name` expressions that reference a name _before_ it's bound via `global` statement -- that's a syntax error:

```py
def f():
  print(a)
  global a
```

The way we do this today is that we iterate to find all globals when we first see the function (this isn't _too_ expensive, since we only have to iterate over statements, and don't need to recurse into other functions or classes), store them in the scope, then as we do the standard traversal, we check each `Expr::Name` against the global list.

An alternative, which wouldn't require us to store these, would be to do a full, dedicated traversal of the function body (if we find any globals during the "find globals" phase), up to the location of the last `global` statement, and check all names. That is: rather than finding the globals upfront, and then checking against names as we iterate over the rest of the AST, we'd find the globals upfront, then find all names upfront too if needed.

In theory, this should be slower, though in practice, it may not matter? Since the vast majority of scopes don't have globals anyway.

Reply:
What's the motivation for removing it? To reduce the size of `Scope`? One thing that we could do is to store the hash map out of bounds. Meaning, we only store a `Option<GlobalsId>` here (4 bytes vs 48 bytes) and store the globals in a `Vec` on scopes (indexed by `GlobalsId`). This adds one additional indirection (or we can store an `Option<Box<HashMap>>`.

Reply:
Yeah, to reduce the size of `Scope`. That's an interesting idea, I'll try it.

Reply:
(Will explore in a separate PR.)